contributing.md

```diff
@@ -0,0 +1,105 @@
+```
+git rebase -i $starting_hash^
+# etc
+```
+
+## github issues
+* include `nim --version` and OS information in PR
+TODO: we should have a cmd line to get all context we need, eg: `nim --issueContext`, analog to `brew gist-logs`
+* make sure to include verbatim relevant parts of error message, to make it easier to search for dedupping
+
+## forum
+### forum title should be descriptive
+Here are some examples take from https://forum.nim-lang.org/
+
+good:
+* Jester v0.3.0 and our first CVE ID
+* How to pass module and function name to call in a template (or macro)?
+
+bad:
+* "explain this behavior for me"
+* "i have some question !?"
+
+## documentation
+exported functions should have nimdoc comments (eg with `##`)
+
+## testing
+
+### runnableExamples
+* try to use `runnableExamples:` block to make generated docs more illustrative and for testing purposes
+
+### unit tests
+* more extensive testing (eg that would be overkill for docs) can be done in `when isMainModule` blocks right below the proc:
+
+```nim
+proc absolutePath*(path: string, root = getCurrentDir()): string =
+    ## Returns the absolute path of `path`, rooted at `root` (which must be absolute)
+    ## if `path` is absolute, return it, ignoring `root`
+    runnableExamples:
+      doAssert absolutePath("a") == getCurrentDir() / "a"
+    if isAbsolute(path): path
+    ...
+  
+  when isMainModule:
+    doAssertRaises(ValueError): discard absolutePath("a", "b")
+    doAssert absolutePath("a") == getCurrentDir() / "a"
+    ...
+```
+
+### integration tests
+TODO
+
+## code
+
+### prefer `proc` > `template` > `macro` when possible
+As general rule of thumb, a `proc` should be preferred over a `template`, and a `template` should be preferred over a `macro` whenever possible (use the simplest tool for the job).
+
+rationale:
+* procedures are more hygienic and easier to debug
+* procedures are compiled only once (or once per instantiating type for generics), unlike templates and macros which are evaluated at every call site; this can affect compile time performance
+* templates and macros don't appear in stackraces, making debugging harder
+* at high enough optimization levels, the C/C++ compiler will inline the corresponding proc anyways so shouldn't affect performance
+
+Caveats with templates and macros: `untyped` parameters can prevent method call syntax in some cases, eg `myIter.toSeq` see [docs](https://nim-lang.org/docs/manual.html#templates-limitations-of-the-method-call-syntax)
+
+### prefer `any` or `void` (or even `auto`) over `untyped` for `template` and `macro` output when possible
+rationale: early type checking makes it easier to debug errors
+
+### prefer `any` or a explicit or generic type over `untyped` for `template` and `macro` parameters when possible
+rationale: early type checking makes it easier to debug errors
+
+###  use `void` instead of `typed` for `template` and `macro` output
```

Comment:
I don't think this suggestion is very clear. `void` means  - well, no return type. As such, it can only be used when the template does not return anything, or so I think. If I have, say, `template foo(): untyped = 5`, I should change that `untyped` into `int`, not `void`